### ISSUE TITLE
dart2native compiler broken in flutter SDK

### ISSUE THREAD

**Issue by mit-mit** [flutter/flutter#43968]
The compiler fails when compiling any program:

```console
$ ~/dev/flutter/bin/cache/dart-sdk/bin/dart2native hello.dart -o ./hello
Failed to generate native files:
FileSystemException: Cannot retrieve length of file, path = '/Users/mit/dev/flutter/bin/cache/dart-sdk/bin/dartaotruntime' (OS Error: No such file or directory, errno = 2)
```

Looks like the problem is that we're not bundling `dartaotruntime`:
```console
$ cd ~/dev/flutter/bin/cache/dart-sdk/bin/
| ~/dev/flutter/bin/cache/dart-sdk/bin @ mit-macbookpro3 (mit)
$ ll
total 56016
-rwxr-xr-x   1 mit  primarygroup    27M Nov  1 00:06 dart
-rwxr-xr-x   1 mit  primarygroup   1.5K Oct 11 19:48 dart2js
-rwxr-xr-x   1 mit  primarygroup   781B Oct 11 19:48 dart2native
-rwxr-xr-x   1 mit  primarygroup   953B Oct 11 19:48 dartanalyzer
-rwxr-xr-x   1 mit  primarygroup   1.1K Oct 11 19:48 dartdevc
-rwxr-xr-x   1 mit  primarygroup   857B Oct 11 19:48 dartdoc
-rwxr-xr-x   1 mit  primarygroup   915B Oct 11 19:48 dartfmt
```

**Comment by mit-mit** [comment]
Tagging TODAY, this totally breaks the feature

**Comment by mit-mit** [comment]
cl/123724 has work on the Dart-side for this, but we still need changes to the Flutter SDK build step

**Comment by mkustermann** [comment]
The agreement was to remove `dart2native` from the flutter-bundled `dart-sdk`. The CL was now submitted and once it (auto)rolls into the flutter/engine and flutter/flutter the `dart2native` script should disappear.

If we decide that we also want the `dart2native` in the flutter-distributed dart-sdk, then there is additional work to be done in the flutter engine build scripts.

**Comment by mit-mit** [comment]
I should clarify: For the immediate need (Dart 2.6, next 1-2 days) the removal is fine, and indeed what we agreed to.

But, for the next set of stable releases (Dart and Flutter) we need the tool to be re-added in the flutter bundled SDK. It's my understanding that means it has to be fixed very soon. TODAY may be the wrong label for that, not sure if we have a "for next stable release" label.

**Comment by zanderso** [comment]
This doesn't look like a TODAY issue, so I've removed the label. Instead, I've added a milestone that indicates that the item is past-due. If this issue will directly affect a customer of the next stable release, please add a 'customer: blocker' or 'customer: critical' label.

**Comment by mkustermann** [comment]
The roll of this from dart into engine happened in https://github.com/flutter/engine/commit/bf9d2633fab5d48e7f6bae68ce30a2dbfdfb52ed . 

I have looked at the engine artifacts uploaded and can confirm that

* [gs://flutter_infra/flutter/ebab0d330b409880b1f52c0c4ed7714020b6b5a3/dart-sdk-linux-x64.zip](https://storage.cloud.google.com/flutter_infra/flutter/ebab0d330b409880b1f52c0c4ed7714020b6b5a3/dart-sdk-linux-x64.zip) - before, has `dart-sdk/bin/dart2native`
* [gs://flutter_infra/flutter/bf9d2633fab5d48e7f6bae68ce30a2dbfdfb52ed/dart-sdk-linux-x64.zip](https://storage.cloud.google.com/flutter_infra/flutter/bf9d2633fab5d48e7f6bae68ce30a2dbfdfb52ed/dart-sdk-linux-x64.zip) - after, does not have `dart-sdk/bin/dart2native` (and is 50 MB smaller)

So once this engine lands in users hands the issue should be gone.

@mit-mit If you want to have a working `bin/dart2native` in the flutter-built sdk, please find an owner from the flutter engine team for this.

What would need to happen is
* The `include_dart2native` GN argument needs to be set in the engine build - similar to our [tools/gn.py](https://github.com/dart-lang/sdk/blob/48ebcdac5f697/tools/gn.py#L249) (that should ensure a working `dart2native`)
* An extra X64-Product build of `gen_snapshot` and `dart_precompiled_runtime` needs to be done and the two binaries need to be copied over the `dart-sdk` built in the previous step, similar to our [tools/bots/dart_sdk.py](https://github.com/dart-lang/sdk/blob/48ebcdac5f697/tools/bots/dart_sdk.py#L74)

**Comment by mit-mit** [comment]
@a-siva can you identify an owner?

**Comment by cbracken** [comment]
@jason-simmons is going to take a look. 

@mit-mit can I understand the push for this? Is the plan to start using `dart2native` in the Dart/Flutter tooling itself? We have never advertised the Dart SDK built into Flutter as being a full Dart SDK that you could use for non-Flutter development, and in fact have gone to great lengths to abstract it away by having the `flutter` tool wrap the underlying `pub` or `dartanalyzer`, etc. commands.

**Comment by cbracken** [comment]
I've moved to the Unassigned customer work milestone, since the engine team is focused on _customer: blocker_ and _customer: critical_ bugs right now.  If we have P0/P1 customers whose launches are blocked on this, please add the appropriate customer label and the critical/blocker label to help prioritise.

**Comment by magillus** [comment]
If Dart SDK bundled with Flutter is different version from normal release, any best practices to have them separate on development machine?
Maybe some option to `flutter upgrade --dart-sdk=bundled` (default) or `--dart-sdk=os` so it would disable bundled Dart sdk and would use the OS provided. 

Maybe a switch so flutter framework would remember user's settings on that machine:
`flutter dart=bundled` or `flutter dart=os`

I would love tools/cli commands to easily upgrade dart sdk based on flutter version that it requires - when ever it would be `bundled` or `os`

**Comment by Hixie** [comment]
Flutter entirely manages the version of Dart it uses, but does not expose it for general use currently.
If you want to use Dart for command line purposes, the way that's supposed to work is to download the Dart SDK separately.

There shouldn't be any risk of them conflicting.

**Comment by bsutton** [comment]
Maybe I'm using it wrong but I'm having troubles with the flutter sdk conflicting with the OS sdk.

I have a dart cli application that is using 'extensions' from 2.6. 

https://pub.dev/packages/dshell

I can dev with vscode fine, as well as publish the code using pub publish.
But if I try into globally activate the package from the dart package repo, the install fails:

```
pub global activate dshell
Package dshell is currently active at path "/home/xxx/dshell".
Resolving dependencies... (1.1s)
The current Dart SDK version is 2.5.0.

Because pub global activate depends on dshell any which requires SDK version >=2.6.0 <3.0.0, version
  solving failed.
```

My OS sdk is:
```
dart --version
Dart VM version: 2.6.1 (Unknown timestamp) on "linux_x64"

```
flutter doctor -v
[✓] Flutter (Channel stable, v1.9.1+hotfix.6, on Linux, locale en_AU.UTF-8)
    • Flutter version 1.9.1+hotfix.6 at /home/xxx/apps/flutter
    • Framework revision 68587a0916 (10 weeks ago), 2019-09-13 19:46:58 -0700
    • Engine revision b863200c37
    • Dart version 2.5.0
```
Looking at my paths I somehow ended up with the fluter dart-sdk on my path:

```
export PATH="$PATH:$HOME/.cargo/bin:/home/xxxx/apps/flutter/bin/cache/dart-sdk/bin:/home/xxxx/apps/flutter/bin"
```

I'm not certain how I got in this state, but it would have been some instruction on the web somewhere telling my to include it.

Having a duplicate copy of the dart tools feels problematic.

**Comment by mit-mit** [comment]
Now that we're exposing the Dart SDK (see https://github.com/flutter/flutter/pull/53470), can we take a look at this again?

The needed steps are:
  1. [Build the product mode files](https://github.com/dart-lang/sdk/blob/3d53df52afb56e797255125e3b038112ce6395be/tools/bots/test_matrix.json#L2800) 
  1. [Copy them into the release SDK](https://github.com/dart-lang/sdk/blob/master/tools/bots/dart_sdk.py#L77)

cc @zanderso

**Comment by zanderso** [comment]
@jason-simmons Could you summarize what the difficulties you ran into were?

I took a cursory look and it seems like we need to set the GN arg `include_dart2native` when the engine gn script is passed `--full-dart-sdk`:

https://github.com/flutter/engine/blob/fa0e464a710790c6110d541ae022d70e2f9009d8/tools/gn#L257

(`include_dart2native` gets plumbed to [here](https://github.com/dart-lang/sdk/blob/master/sdk/BUILD.gn#L25))

Does the engine Dart SDK builder then apply any filters during packaging? Maybe @dnfield knows.

**Comment by dnfield** [comment]
https://flutter.googlesource.com/recipes/+/refs/heads/master/recipes/engine.py#363 - whatever we build for the SDK should get uploaded without any extra steps.

**Comment by zanderso** [comment]
@jason-simmons has informed me that I've underestimated the complexity of this. If we're going to do the exact same thing as the Dart team does, we'd have to build `dart2native` in product mode and add extra steps to the recipes to manually copy the product mode binaries into the release mode SDK.

Unless my knowledge of the Dart GN build is out of date, I think this could have been accomplished by the Dart team entirely in the GN build rather than using an out-of-band packaging step. In particular, binaries can be hard-coded to product mode even in a release build. See `gen_snapshot_product` here for an example:

https://github.com/dart-lang/sdk/blob/master/runtime/bin/BUILD.gn#L239

**Comment by dnfield** [comment]
Why aren't we using product mode in release?

**Comment by jonahwilliams** [comment]
product mode removes the vm service and means Dart VM applications couldn't be debugged

**Comment by dnfield** [comment]
But don't we do that anyway for release?

**Comment by jonahwilliams** [comment]
you're talking about flutter, this is for `bin/cache/dart-sdk` which we use for all of our tooling

**Comment by dnfield** [comment]
Ahhhh got it.

**Comment by mkustermann** [comment]
> Unless my knowledge of the Dart GN build is out of date, I think this could have been 
> accomplished by the Dart team entirely in the GN build rather than using an out-of-band 
> packaging step. In particular, binaries can be hard-coded to product mode even in a release 
> build. See gen_snapshot_product here for an example:

@zanderso Yes that would be the cleanest (I wonder why 907c514c8937cf76e825d427364cf3475797f18f was done in the way it was)

**Comment by zanderso** [comment]
I filed https://github.com/dart-lang/sdk/issues/42230

**Comment by zanderso** [comment]
I note that working `dart2native` and `dartaotruntime` are now in the Flutter SDK on master.

**Comment by abulka** [comment]
I'm on the flutter sdk beta channel and cannot see `dart2native` yet - presumably being on master doesn't necessarily mean being in beta. 
Hopefully a new beta release will happen with `dart2native`  in it soon?

**Comment by mit-mit** [comment]
Hi @abulka, yes the current beta build is from June 9th, which it prior to this change landing in master:

```
$ bin/dart --version
Dart VM version: 2.9.0-14.1.beta (beta) (Tue Jun 9 10:52:57 2020 +0200) on "macos_x64"
```

**Comment by Timmmm** [comment]
I am using Flutter on the beta channel and don't seem to have `dart2native`. What am I missing?

```sh
$ flutter --version
Flutter 1.25.0-8.1.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision 8f89f6505b (11 days ago) • 2020-12-15 15:07:52 -0800
Engine • revision 92ae191c17
Tools • Dart 2.12.0 (build 2.12.0-133.2.beta)
$ which flutter
/Users/me/local/flutter/bin/flutter
$ ls ~/local/flutter/bin 
cache       dart        dart.bat    flutter     flutter.bat internal
```

**Comment by bsutton** [comment]
The compulsory is now in the dart exe.

Use dart compile exe <library>
